Lib/profiling/sampling/heatmap_collector.py

```diff
@@ -0,0 +1,1038 @@
+        code_lines_html = [
+            self._build_line_html(line_num, line_content, line_counts, self_counts,
+                                max_samples, max_self_samples, filename)
+            for line_num, line_content in enumerate(source_lines, start=1)
+        ]
+
+        # Populate template
+        replacements = {
+            "<!-- FILENAME -->": html.escape(filename),
+            "<!-- TOTAL_SAMPLES -->": f"{file_stat.total_samples:,}",
+            "<!-- TOTAL_SELF_SAMPLES -->": f"{file_stat.total_self_samples:,}",
+            "<!-- NUM_LINES -->": str(file_stat.num_lines),
+            "<!-- PERCENTAGE -->": f"{file_stat.percentage:.2f}",
+            "<!-- MAX_SAMPLES -->": str(file_stat.max_samples),
+            "<!-- MAX_SELF_SAMPLES -->": str(file_stat.max_self_samples),
+            "<!-- CODE_LINES -->": ''.join(code_lines_html),
+            "<!-- INLINE_CSS -->": f"<style>\n{self._template_loader.file_css}\n</style>",
+            "<!-- INLINE_JS -->": f"<script>\n{self._template_loader.file_js}\n</script>",
+        }
+
+        html_content = self._template_loader.file_template
+        for placeholder, value in replacements.items():
+            html_content = html_content.replace(placeholder, value)
+
+        try:
+            output_path.write_text(html_content, encoding='utf-8')
+        except (IOError, OSError) as e:
+            raise RuntimeError(f"Failed to write file {output_path}: {e}") from e
+
+    def _build_line_html(self, line_num: int, line_content: str,
+                        line_counts: Dict[int, int], self_counts: Dict[int, int],
+                        max_samples: int, max_self_samples: int, filename: str) -> str:
+        """Build HTML for a single line of source code."""
+        cumulative_samples = line_counts.get(line_num, 0)
+        self_samples = self_counts.get(line_num, 0)
+
+        # Calculate colors for both self and cumulative modes
+        if cumulative_samples > 0:
+            cumulative_intensity = cumulative_samples / max_samples if max_samples > 0 else 0
+            self_intensity = self_samples / max_self_samples if max_self_samples > 0 and self_samples > 0 else 0
+
+            # Default to self-based coloring
+            intensity = self_intensity if self_samples > 0 else cumulative_intensity
+            r, g, b, alpha = self._calculate_intensity_color(intensity)
+            bg_color = f"rgba({r}, {g}, {b}, {alpha})"
+
+            # Pre-calculate colors for both modes (for JS toggle)
+            self_bg_color = self._format_color_for_intensity(self_intensity) if self_samples > 0 else "transparent"
+            cumulative_bg_color = self._format_color_for_intensity(cumulative_intensity)
+
+            self_display = f"{self_samples:,}" if self_samples > 0 else ""
+            cumulative_display = f"{cumulative_samples:,}"
+            tooltip = f"Self: {self_samples:,}, Cumulative: {cumulative_samples:,}"
```

Comment:
Should we call this _Total_ to stay consistent with the table, since cumulative values are displayed as _Total_ there
```suggestion
            tooltip = f"Self: {self_samples:,}, Total: {cumulative_samples:,}"
```